### ISSUE TITLE
Federation catchup for a destination can block indefinitely on un-partial stating

### ISSUE THREAD

**Issue by squahtx** [matrix-org/synapse#15220]
**What we think happened**
1. A user faster joined a remote room
2. Some events were sent into the room, but not transmitted successfully
3. The user left the room before it became full state
4. The federation catchup loop tried to transmit events in the room, but blocked on full state. This also blocked \*all\* catchup transmission for the destination.

**Stack trace**
```
  File "/home/synapse/src/synapse/metrics/background_process_metrics.py", line 242, in run
    return await func(*args, **kwargs)
  File "/home/synapse/src/synapse/federation/sender/per_destination_queue.py", line 321, in _transaction_transmission_loop
    await self._catch_up_transmission_loop()
  File "/home/synapse/src/synapse/federation/sender/per_destination_queue.py", line 544, in _catch_up_transmission_loop
    new_pdus = await filter_events_for_server(
  File "/home/synapse/src/synapse/visibility.py", line 650, in filter_events_for_server
    event_to_memberships = await _event_to_memberships(
  File "/home/synapse/src/synapse/visibility.py", line 746, in _event_to_memberships
    event_to_state_ids = await storage.state.get_state_ids_for_events(
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/controllers/state.py", line 267, in get_state_ids_for_events
    event_to_groups = await self.get_state_group_for_events(
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/controllers/state.py", line 374, in get_state_group_for_events
    await self._partial_state_events_tracker.await_full_state(event_ids)
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/util/partial_state_events_tracker.py", line 82, in await_full_state
    logger.info(
```

**Comment by DMRobertson** [comment]
Workedaround in https://github.com/matrix-org/synapse/commit/ab629c17ccfbc98e8297e7fd3fcf03b641bd3cc0, but I think we ought to fix this properly for the next release.

**Comment by DMRobertson** [comment]
Of note: we only reach that far down into the stack if the room's history visibility is restricted:

https://github.com/matrix-org/synapse/blob/b5b5f6608462a988b05502a3b70b6a57ca3846d2/synapse/visibility.py#L649-L676

**Comment by DMRobertson** [comment]
Fixed by #15248.
